filament/include/filament/BufferObject.h

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! \file
+
+#ifndef TNT_FILAMENT_BUFFEROBJECT_H
+#define TNT_FILAMENT_BUFFEROBJECT_H
+
+#include <filament/FilamentAPI.h>
+
+#include <backend/DriverEnums.h>
+
+#include <backend/BufferDescriptor.h>
+
+#include <utils/compiler.h>
+
+namespace filament {
+
+class FBufferObject;
+
+class Engine;
+
+/**
+ * A generic GPU buffer containing data.
+ *
+ * Usage of this BufferObject is optional. For simple use cases it is not necessary. It is useful
+ * only when you need to share data between multiple VertexBuffer instances. It also allows you to
+ * efficiently swap-out the buffers in VertexBuffer.
+ *
+ * NOTE: For now this is only used for vertex data, but in the future we may use it for other things
+ * (e.g. compute).
+ *
+ * @see VertexBuffer
+ */
+class UTILS_PUBLIC BufferObject : public FilamentAPI {
+    struct BuilderDetails;
+
+public:
+    using BufferDescriptor = backend::BufferDescriptor;
+
+    class Builder : public BuilderBase<BuilderDetails> {
+        friend struct BuilderDetails;
+    public:
+        Builder() noexcept;
+        Builder(Builder const& rhs) noexcept;
+        Builder(Builder&& rhs) noexcept;
+        ~Builder() noexcept;
+        Builder& operator=(Builder const& rhs) noexcept;
+        Builder& operator=(Builder&& rhs) noexcept;
+
+        /**
+         * Size of the buffer in bytes.
+         * @param byteCount Maximum number of bytes the BufferObject can hold.
+         * @return A reference to this Builder for chaining calls.
+         */
+        Builder& size(uint32_t byteCount) noexcept;
```

Comment:
Do we need a `.type()` method to distinguish between SSBO, VBO, etc?
Maybe we should add it now with only a single value possible and documenting it defaults to VBO?

Reply:
Yes I love this idea.  Perhaps it could be called `bindingTarget`.

I suppose we could even support two of these for now: STATIC_VERTEX and DYNAMIC_VERTEX.  However I'm really not sure if GL's STATIC vs DYNAMIC hint actually does anything...